### ISSUE TITLE
Documentation request: deep linking and Android launchMode

### ISSUE THREAD

**Issue by olof-dev** [flutter/flutter#110928]
tl;dr: Please could a knowledgeable Android dev (1) describe in [the deep linking docs](https://docs.flutter.dev/development/ui/navigation/deep-linking) how to prevent a second copy of the app being launched when a deep link is tapped, and (2) document the (surprising) default behavior.

## The issue

With deep linking into an Android Flutter app enabled as per [the docs](https://docs.flutter.dev/development/ui/navigation/deep-linking), with no other changes to AndroidManifest.xml, the default behavior on several Android versions (e.g. 6, 7, 9) is for a tap on a deep link to open a *second* copy of the Flutter app, if one is already running when the link is tapped. This wasn't obvious, and in our app caused very hard-to-diagnose problems with Firebase authentication. (It seems that on Android 12, the deep link opens in the already-running app. I haven't tested 10 or 11.)

## Proposal

Firstly, it'd be great if this behavior was documented in [the deep linking docs](https://docs.flutter.dev/development/ui/navigation/deep-linking).

Secondly, it'd be great if a workaround could be described. Here's what I've managed to surmise (but I have little Android dev experience):

The behavior seems to be related to the setting `android:launchMode` in `.MainActivity` in AndroidManifest.xml, which in Flutter projects defaults to `singleTop`. In a number of existing issues/PRs, and on StackOverflow, there are suggestions to change this to `singleTask`; see for example

https://github.com/flutter/flutter/pull/56899
https://github.com/flutter/flutter/issues/91649
https://github.com/flutter/samples/pull/883

But, as linked to in the first issue, using `singleTask` [apparently](https://blog.dixitaditya.com/android-task-hijacking/amp/) opens the app up to "task highjacking", and is also discouraged by [the Android docs on launchMode](https://developer.android.com/guide/topics/manifest/activity-element#lmode).

Using `singleInstance` guarantees that only a single instance of the app runs, and seems to work with deep linking, and not be susceptible to task highjacking, but maybe there are drawbacks to using it, even in a Flutter environment? Its use is also discouraged by the Android docs for some reason.

Perhaps a different solution might be to come from the `intent`-side, and use the `FLAG_ACTIVITY_*` constants, but I don't know how they fit with Flutter.

Some guidance about this in the docs would be great!

**Comment by yanshouwang** [comment]
The launchMode config really should be addressed in the [deep link](https://docs.flutter.dev/ui/navigation/deep-linking) or [app link](https://docs.flutter.dev/cookbook/navigation/set-up-app-links) docs.

**Comment by matt-lethargic** [comment]
Is there any movement or advice on this? The application I'm developing just opens a new instance in whatever email client I use and then breaks the rest of the functionality. :-(

**Comment by haykgalstyan** [comment]
Same. SingleTop was causing new instance of my up opened on every deep link redirect. I found out the solutions after wasting a couple of days... This should at-least be documented in the deep link docs.
I also ended up here wondering why singleTask was not the default, and whether my change from singleTop to singleTask will break some other stuff in flutter.

**Comment by orkun1675** [comment]
@GaryQian can we reconsider the priority for this bug? I've been working with Flutter for 2+ years and this is an area that is still opaque to me.

The default launch modes (`standard` and `singleTop`) also break some plugins after a deep-link-click, causing platform channels to stop working. It's super hard to debug these without understanding the inner workings of the Flutter activity and engine lifecycle.

**Comment by utopicnarwhal** [comment]
I found an interesting workaround. Maybe someone would like to test it: https://support.mocaplatform.com/portal/en/kb/articles/recommendation-for-deep-links-in-android

**Comment by utopicnarwhal** [comment]
Seems like it works.

Sharing the code that I used:
```kotlin
// android/app/src/main/kotlin/com/example/app/IntentForwardingActivity.kt
package com.example.app

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity

/**
 * Acts as a singleTask root for deep links into the app.
 *
 * Declare this Activity in the manifest with:
 *     launchMode="singleTask"
 *
 * This lets deep links be handled in a separate task from the host app
 * without forcing LaunchActivity itself to be singleTask (which would
 * wipe its back stack). It also makes it easy to clear the stack when
 * exiting or restarting the app.
 */
class IntentForwardingActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        handleIntent(intent)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
        handleIntent(intent)
    }

    private fun handleIntent(incoming: Intent?) {
        // Prepare the intent to forward to MainActivity
        // Ensure that MainActivity is set to launchMode="singleTop" in your AndroidManifest.xml
        val forwardIntent = Intent(this, MainActivity::class.java).apply {
            // Copy data and extras from the incoming intent
            action = incoming?.action
            data = incoming?.data
            if (incoming?.extras != null) {
                putExtras(incoming.extras!!)
            }
            flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
            // Add any flags from the original intent, or set new ones as needed
            // For example, if you want MainActivity to be at the top of the stack
            // and receive the intent via onNewIntent if it's already running:
            // flags = Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_CLEAR_TOP
            // However, since MainActivity is already singleTop, FLAG_ACTIVITY_SINGLE_TOP is inherent.
            // FLAG_ACTIVITY_CLEAR_TOP could be used if you want to clear activities above MainActivity
            // in its task.
        }

        startActivity(forwardIntent)
        finish() //  Close this IntentForwardingActivity after forwarding the intent
    }
}
```

```xml
// AndroidManifest.xml

<manifest xmlns:tools="http://schemas.android.com/tools"
    xmlns:android="http://schemas.android.com/apk/res/android">

    ...

    <application
        android:label="@string/app_name"
        android:name="${applicationName}"
        android:localeConfig="@xml/locales_config"
        android:dataExtractionRules="@xml/data_extraction_rules"
        android:allowBackup="false"
        android:fullBackupContent="false"
        android:enableOnBackInvokedCallback="true"
        tools:targetApi="tiramisu"
        android:icon="@mipmap/launch_icon"
        android:roundIcon="@mipmap/launch_icon_round">
        <activity
            android:name=".MainActivity"
            android:exported="true"
            android:launchMode="singleTop"
            android:theme="@style/LaunchTheme"
            android:configChanges="orientation|keyboardHidden|keyboard|screenSize|smallestScreenSize|locale|layoutDirection|fontScale|screenLayout|density|uiMode"
            android:hardwareAccelerated="true"
            android:windowSoftInputMode="adjustResize">
            <meta-data
              android:name="io.flutter.embedding.android.NormalTheme"
              android:resource="@style/NormalTheme"
              />
            <intent-filter>
                <action android:name="android.intent.action.MAIN"/>
                <category android:name="android.intent.category.LAUNCHER"/>
            </intent-filter>

            <!-- App links setup -->
            <meta-data
                android:name="flutter_deeplinking_enabled"
                android:value="false" />
        </activity>

        <!-- Handle redirect in the app instead of the web -->
        <!-- 
            Because we have to use the "singleTop" main activity launch mode — any time a deep link is clicked
            we get a second instance of the app. That instance of your activity will open in the app that hosts the link.
            Then there can be two (potentially) full-fledged versions of an app running in two different tasks.

            Source: https://support.mocaplatform.com/portal/en/kb/articles/recommendation-for-deep-links-in-android
        -->
        <activity android:name=".IntentForwardingActivity"
            android:launchMode="singleTask"
            android:exported="true">
            <intent-filter android:autoVerify="true">
                <action android:name="android.intent.action.VIEW" />
                <category android:name="android.intent.category.DEFAULT" />
                <category android:name="android.intent.category.BROWSABLE" />
                <data android:scheme="https" />
                <data android:scheme="http" />
                <data android:host="@string/oauth2_redirect_domain" />
                <data android:pathPrefix="/api/oauth2-redirect" />
            </intent-filter>

            <!-- Just in case -->
            <meta-data
                android:name="flutter_deeplinking_enabled"
                android:value="false" />
        </activity>

        <meta-data
            android:name="io.flutter.embedding.android.EnableImpeller"
            android:value="false" />
        <meta-data
            android:name="com.google.firebase.messaging.default_notification_icon"
            android:resource="@drawable/ic_notification" />
        <meta-data
            android:name="flutterEmbedding"
            android:value="2" />
    </application>
</manifest>
```

Would be nice if someone else could test it and share the results.

**Comment by yasargil** [comment]
@utopicnarwhal Looks like a nice solution, will try that with our next version. 

@Flutter Team. Changing the launchMode from singleTop to singleTask did causes problems for us with external native activities. Because that launchMode discard all activities above the main one.

**Comment by maoja37** [comment]
@utopicnarwhal Just tried it
Didn't work for unfortunately

I still have the problem where it opens a new instance of the app every time and kind of breaks the flow of the app

This isn't a problem on ios tho, I wonder why it is a problem on Android

**Comment by utopicnarwhal** [comment]
> @utopicnarwhal Just tried it
> Didn't work for unfortunately
> 
> I still have the problem where it opens a new instance of the app every time and kind of breaks the flow of the app
> 
> This isn't a problem on ios tho, I wonder why it is a problem on Android

Have you tried to install the app without flutter debug mode? Idk why but it doesn't work when it's run through Flutter run
